mindmeld/auto_annotator.py

```diff
@@ -0,0 +1,708 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2015 Cisco Systems, Inc. and others.  All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABC, abstractmethod
+import re
+import logging
+import os
+import importlib
+from enum import Enum
+from tqdm import tqdm
+import spacy
+
+
+from .resource_loader import ResourceLoader
+from .components._config import get_auto_annotator_config
+from .system_entity_recognizer import DucklingRecognizer
+from .markup import load_query, dump_queries
+from .core import Entity, Span, QueryEntity
+from .query_factory import QueryFactory
+from .exceptions import MarkupError
+from .models.helpers import register_annotator
+
+logger = logging.getLogger(__name__)
+
+
+class AnnotatorAction(Enum):
+    ANNOTATE = "annotate"
+    UNANNOTATE = "unannotate"
+
+
+class Annotator(ABC):
+    """
+    Abstract Annotator class that can be used to build a custom Annotation class.
+    """
+
+    def __init__(self, app_path, config=None, resource_loader=None):
+        """ Initializes an annotator."""
+        self.app_path = app_path
+        self.config = get_auto_annotator_config(app_path=app_path, config=config)
+        self._resource_loader = (
+            resource_loader or ResourceLoader.create_resource_loader(app_path)
+        )
+        self.annotate_file_entities_map = self._get_file_entities_map(action="annotate")
+
+    def _get_file_entities_map(self, action="annotate"):
```

Comment:
I think this comment by @tmehlinger is quite prescient here: https://sqbu-github.cisco.com/WebexIntelligence/joint_bert_model/pull/1

Let's state types for every function using mypy in the function arguments so that we dont have to do these types of checks:
```
        if action == AnnotatorAction.ANNOTATE.value:
            rules = self.config[AnnotatorAction.ANNOTATE.value]
        elif action == AnnotatorAction.UNANNOTATE.value:
            rules = self.config[AnnotatorAction.UNANNOTATE.value]
```

Instead, we can do this:
```
        if action == AnnotatorAction.ANNOTATE:
            rules = self.config[AnnotatorAction.ANNOTATE]
        elif action == AnnotatorAction.UNANNOTATE:
            rules = self.config[AnnotatorAction.UNANNOTATE]
```

Reply:
Let's not annotate this whole PR with types. But let's use it for functions like these where enums are expected.

Reply:
Could you please elaborate on how we can use mypy to be able to do the second type of checks?

Currently, I am getting an error when comparing a string with an enum value as in the case of the second set of checks above

Reply:
Updated to use mypy. The function now takes AnnotatorAction as the type for the action parameter.

Needed to use the value in order to index into the config dict with a str

`rules = self.config[AnnotatorAction.ANNOTATE.value]`